cloudinit/sources/__init__.py

```diff
@@ -408,6 +411,16 @@ def persist_instance_data(self):
         json_sensitive_file = os.path.join(
             self.paths.run_dir, INSTANCE_JSON_SENSITIVE_FILE
         )
+        cloud_id = instance_data["v1"].get("cloud_id", "none")
+        cloud_id_file = os.path.join(self.paths.run_dir, "cloud-id")
+        util.write_file(f"{cloud_id_file}-{cloud_id}", cloud_id)
+        if os.path.exists(cloud_id_file):
+            prev_cloud_id_file = os.path.realpath(cloud_id_file)
+        else:
+            prev_cloud_id_file = cloud_id_file
+        util.sym_link(f"{cloud_id_file}-{cloud_id}", cloud_id_file, force=True)
+        if prev_cloud_id_file != cloud_id_file:
+            util.del_file(prev_cloud_id_file)
```

Comment:
Is this necessary? `/run/cloud-init` gets removed every boot. Do we support changing a datasource post-boot? It doesn't hurt anything though.

Reply:
I know of some AWS customers whose upgrade and deployment strategy involves executing `cloud-init clean --logs; cloud-init init --locl; cloud-init init`. If someone injected NoCloud seed files before upgrade I could see a stale /run/cloud-init/cloud-id-ec2 left around instead of  /run/cloud-init/cloud-id-no-cloud. Just didn't want occasion to have artifactts around that are no longer correct. There also may be folks that live-migrate a VM to a sandboxed environment which may no longer have visibility to the original IMDS. If they trigger a persist_instance_data() we'd want to cleanup any no longer applicable DS in case something else reacts to it's presence.